### DISCUSSION TITLE
How to use puppeteer?

### DISCUSSION THREAD

**Discussion by srcnalt** [vercel/community#124] [Show & Tell]
Hi, I'm trying to get puppeteer working on vercel, does anyone have experience getting it working?

I get the following error at the point browser is launched:
`puppeteer failed: Error: Could not find browser revision 768783. Run "PUPPETEER_PRODUCT=firefox npm install" or "PUPPETEER_PRODUCT=firefox yarn install" to download a supported Firefox browser binary.`

It should be running chromium, but the error suggests using firefox, which I could not find a way how to do.

**Comment by srcnalt** [comment]
I ended up using Playwright.

**Reply by lewyuburi** [reply]
Playwright works right with Vercel?

**Reply by srcnalt** [reply]
Yes it worked just fine for me, does everything Puppeteer does and some more.

**Reply by metinferatii** [reply]
@srcnalt I couldn't make Playwright on Vercel it gives me the "Failed to launch chromium because executable doesn't exist at" error. Did you face any problems like this?

**Reply by srcnalt** [reply]
I had these in package.json, could you give it a try? I cant remember exactly but, there was some version match issues. So I had to use theme like this.

```
"playwright-aws-lambda": "^0.5.2"
"playwright-core": "npm:playwright-chromium@1.1.1"
```

**Reply by rfoel** [reply]
I'm using vercel without next.js and it keeps giving me this:
```
2021-02-18T18:28:13.124Z	undefined	ERROR	Cannot find module 'playwright'
Require stack:
- /var/task/api/sync.ts
- /var/task/___vc_launcher.js
- /var/runtime/UserFunction.js
- /var/runtime/index.js
2021-02-18T18:28:13.125Z	undefined	ERROR	Did you forget to add it to "dependencies" in `package.json`?
RequestId: 384c6319-cdd3-43cb-8b30-3d4d4a18b23a Error: Runtime exited with error: exit status 1
Runtime.ExitError
```
Any idea why my dependencies are not being carried on deploy? Do you have an example repo of it working with either playwright or puppeteer?

**Reply by vasco3** [reply]
@rfoel are you importing it or are you using `require()`?

**Reply by metinferatii** [reply]
@rfoel @vasco3 no one works with the free one because it limits installing/importing chromium, I just ended up switching to Digital Ocean

**Reply by vasco3** [reply]
@srcnalt do you use it for pdf generation? I am using puppeteer but for pdfs it splits the pages without respecting page breaks so it slices text and images

**Reply by vasco3** [reply]
@psikomanjak ah ok. I use the pro plan

**Reply by ycjcl868** [reply]
browserType.launch: Failed to launch chromium because executable doesn't exist at /vercel/.cache/ms-playwright/chromium-878941/chrome-linux/chrome

**Reply by winston0410** [reply]
@ycjcl868 I have the same error. How did you solved that?

**Reply by ycjcl868** [reply]
I gave up playwright and puppeteer.

**Comment by diegoguevara** [comment]
Hi, this worked for me:

first: 

```
npm add chrome-aws-lambda
```

then:

```js
// import * as puppeteer from 'puppeteer';
import chromium from 'chrome-aws-lambda';

// const browser = await puppeteer.launch( { args: ['--no-sandbox'] } );

const browser = await chromium.puppeteer.launch({
    args: [...chromium.args, "--hide-scrollbars", "--disable-web-security"],
    defaultViewport: chromium.defaultViewport,
    executablePath: await chromium.executablePath,
    headless: true,
    ignoreHTTPSErrors: true,
  })
```

**Reply by 28development** [reply]
Which packages did you install, only ```chrome-aws-lambda```, does this work if you deploy it on vercel, I think I had the issue that vercel was telling me to install chrome. Detailed information would be great :)

**Reply by ndom91** [reply]
You also need `puppeteer-core`/`playwright-core` I believe. This reuses the version of chrome preinstalled in a layer or whatever in vercel function handlers (AWS lambdas).

**Reply by 28development** [reply]
@ndom91 thanks, could you provide an example, not sure where you use playright-core here

**Reply by ndom91** [reply]
> @ndom91 thanks, could you provide an example, not sure where you use playright-core here
> 

Yeah I actually wrote a blog post about it haha 

https://ndo.dev/posts/link-screenshot

**Reply by 28development** [reply]
@ndom91 awesome I am going to try that out, that would literally save my day ❤️

**Reply by 28development** [reply]
@ndom91 so I create an api directory in the root of the project right? currently the api directory is inside pages/api/mystuff.js which is the default location

**Reply by ndom91** [reply]
Yeah so that is an issue with vercel next.js bundling. If you put it in the default location (pages/api/..) the bundling / code sharing will create a bundle that's over 50mb and too big for free vercel accounts. 

Thats why I put it in /api in the root. Let me see if I can find the issue that explained this better than I could haha 

Edit: https://github.com/vercel/vercel/issues/4739

Edit2: if using the root api directory, ` next dev` unfortunately won't pick it up. So you have to use `vercel` cli for dev server from now on :/ more details towards the bottom of that linked issue ^

**Reply by 28development** [reply]
@ndom91 but the 50mb is not limited to the free accounts right? I mean if that the issue I have no problem to update my plan to pro 😆

**Reply by ndom91** [reply]
Yeah that limit is only for free accounts. If you have the budget for paid then you can basically do the same thing as in my code in the blog post, but keep the api route in the default location and continue to be able to use `next dev`

**Reply by 28development** [reply]
@ndom91 do you mind having a chat on discord? I am currently confused about a few things? I am going to write you an email if that is ok

**Reply by amccloud** [reply]
Thanks for the article. The limit for lambda deployments are 50mb free or paid from my experience.

From my understand this is lambda limitation, even if you’re paying Vercel / AWS. 

https://docs.aws.amazon.com/lambda/latest/dg/gettingstarted-limits.html

**Reply by amyegan** [reply]
That's right, it is a Lambda limitation. The 50 MB [Serverless Function size limit](https://vercel.com/docs/concepts/limits/overview#serverless-function-size) is the same across all account types. But things like [Serverless Function Memory](https://vercel.com/docs/concepts/limits/overview#serverless-function-memory) and [Serverless Function Execution Timeout](https://vercel.com/docs/concepts/limits/overview#serverless-function-execution-timeout) do increase on Pro.

**Comment by munkychop** [comment]
@diegoguevara's answer was really helpful (thank you!), but the missing piece for me was that this won't work locally (at least with my basic express app set up), so I thought it didn't work at all, even though it _did_ in fact work when published to Vercel.

My approach to get the best of both worlds (which I partially borrowed from a Japanese article 😄) was to detect whether or not `AWS_LAMBDA_FUNCTION_VERSION` is available and then load an appropriate version of puppeteer:

```
let chrome = {};
let puppeteer;

if (process.env.AWS_LAMBDA_FUNCTION_VERSION) {
  // running on the Vercel platform.
  chrome = require('chrome-aws-lambda');
  puppeteer = require('puppeteer-core');
} else {
  // running locally.
  puppeteer = require('puppeteer');
}

const getData = async (url) => {
  try {
    let browser = await puppeteer.launch({
      args: [...chrome.args, '--hide-scrollbars', '--disable-web-security'],
      defaultViewport: chrome.defaultViewport,
      executablePath: await chrome.executablePath,
      headless: true,
      ignoreHTTPSErrors: true,
    });
  } catch (err) {
    console.error(err);
    return null;
  }
}

```

Note that, because of this, I installed `puppeteer-core` and `chrome-aws-lambda` as dependencies, but installed `puppeteer` as a dev dependency.

**Reply by 28development** [reply]
Did you locally get the error that chrome was not installed aswell? I get the same error deploying this on vercel (chrome not installed). You did not add puppeteer in your packages right, only chrome-aws-lambda and puppeteer-core regarding this feature?

**Reply by munkychop** [reply]
I added `puppeteer` as a dev dependency, so that I could load it locally.

**Reply by stubbzilla8** [reply]
This worked wonderfully!

**Comment by mihhail-lapushkin** [comment]
This should work:
https://github.com/vercel/virtual-event-starter-kit/blob/main/lib/screenshot.ts
```
import chrome from 'chrome-aws-lambda';
import puppeteer from 'puppeteer-core';

export default async function screenshot(url: string) {
  const options = process.env.AWS_REGION
    ? {
        args: chrome.args,
        executablePath: await chrome.executablePath,
        headless: chrome.headless
      }
    : {
        args: [],
        executablePath:
          process.platform === 'win32'
            ? 'C:\\Program Files (x86)\\Google\\Chrome\\Application\\chrome.exe'
            : process.platform === 'linux'
            ? '/usr/bin/google-chrome'
            : '/Applications/Google Chrome.app/Contents/MacOS/Google Chrome'
      };
  const browser = await puppeteer.launch(options);
  const page = await browser.newPage();
  await page.setViewport({ width: 2000, height: 1000 });
  await page.goto(url, { waitUntil: 'networkidle0' });
  return await page.screenshot({ type: 'png' });
}
```

The only thing you need to do, is to manually set either one of these environment variables to some value:
https://github.com/alixaxel/chrome-aws-lambda/blob/master/source/index.ts#L199
```
AWS_LAMBDA_FUNCTION_NAME
FUNCTION_NAME
FUNCTION_TARGET
FUNCTIONS_EMULATOR
```

For some reason, they are not set in Vercel.

h/t @leerob

**Reply by fgiuliani** [reply]
hey @mihhail-lapushkin I tested this and it's not working for me. I'm getting "The Serverless Function "api" is 50.17mb which exceeds the maximum size limit of 50mb". Have you faced this issue before?

**Reply by mihhail-lapushkin** [reply]
Didn't have this problem.

I've tried it on a fresh repo. Worked like a charm.

https://github.com/mihhail-lapushkin/vercel-puppeteer
https://vercel-puppeteer-demo.vercel.app

**Reply by fgiuliani** [reply]
Thank you for the help @mihhail-lapushkin

**Reply by fgiuliani** [reply]
@mihhail-lapushkin I'm still having the issue. I can't find what's the problem. Do you have a minute to take a look at my repo, to see If everything is ok for you?

https://github.com/fgiuliani/automatic-social-images

Again, thank you very much

**Reply by mihhail-lapushkin** [reply]
I think the difference is that I am generating images at build time and you are on every request. The environments for building and running might be different.

I am not very familiar with Next.js, so can't help you there.

**Reply by jgalianoz** [reply]
@mihhail-lapushkin One question, where should the variables be set? Since AWS_LAMBDA_FUNCTION_NAME is a reserved variable in vercel.

**Reply by mihhail-lapushkin** [reply]
I set it in my `package.json` > `scripts` like so:
```
"scripts": {
  "build": "NODE_ENV=production AWS_LAMBDA_FUNCTION_NAME=true <your build command>"
}
```

**Comment by bigsee** [comment]
I also had an issue with this. This works locally for me absolutely fine using `vercel dev`. However, when I deploy this to Vercel, I get the following error:

```
A plugin listed 'puppeteer-extra-plugin-stealth/evasions/chrome.app' as dependency, which is currently missing.
Please install it:  yarn add puppeteer-extra-plugin-stealth          
Note: You don't need to require the plugin yourself, unless you want to modify it's default settings.
``` 

This dependency is in my `package.json`. Any ideas about how to solve this error?

**Reply by aditodkar** [reply]
@bigsee Did you manage to resolve it? I am facing same issue.

**Comment by ndom91** [comment]
If anyone comes across this thread, just wanted to point to this closed issue: https://github.com/vercel/vercel/issues/4739

Basically theres still, at the time of writing, a vercel deployment issue - something about sharing code between the api routes and other files which cause the size to balloon. If you move your `/pages/api/..` directory to simply `/api` then when vercel builds your project, it will come in under the 50mb limit for some reason haha

**Reply by nk1tz** [reply]
@ndom91 This doesn't seem to work. 
Vercel doesn't deploy the `/api` directory the way it does with `/pages/api`.

I'm using `next@11.0.1`.

Could you give more detail about your setup where you achieved this hack?

**Reply by ndom91** [reply]
Hey so as far as i remember thats all I did. You gotta make sure you use `playwright-core` and `chrome-aws-lambda` to keep the size down. 

You can check out the repo in which I did this here - https://github.com/ndom91/home2021. 

Also using next@11+.

**Comment by Julian-quintero** [comment]
Im getting this error after trying every single solution in this thread, any help please.

`error  Error: spawn ETXTBSY
    at ChildProcess.spawn (internal/child_process.js:403:11)
    at spawn (child_process.js:667:9)
    at Object.spawnWithSignal (child_process.js:871:17)
    at BrowserRunner.start (/var/task/node_modules/puppeteer-core/lib/cjs/puppeteer/node/BrowserRunner.js:80:34)
    at ChromeLauncher.launch (/var/task/node_modules/puppeteer-core/lib/cjs/puppeteer/node/Launcher.js:94:16)
    at async funt (/var/task/index.js:33:21) {
  errno: -26,
  code: 'ETXTBSY',
  syscall: 'spawn'
}`

**Reply by vasco3** [reply]
I had to put it on its own verbal project so that it won't run out of memory space.

**Comment by CWarth93** [comment]
I tried everything mentioned here but still cant get it working.

My dependencies:
```
"dependencies": {
    "chrome-aws-lambda": "^10.1.0",
    "dotenv": "^8.2.0",
    "moment": "^2.29.1",
    "next": "10.0.0",
    "prettier": "^2.0.2",
    "puppeteer-core": "^13.5.1",
    "react": "^16.13.1",
    "react-youtube": "^7.14.0",
    "riteway": "^6.1.2"
  },
  "devDependencies": {
    "puppeteer": "^13.5.1"
  }
```
  
 How i try to start the browser:
 ```
import chromium from 'chrome-aws-lambda';

const getBrowser = async () => {
	try { 
		const executablePath = await chromium.executablePath;
		if (executablePath === null) {
			// running locally
			const puppeteer = require('puppeteer');
			return puppeteer.launch({
				headless: true,
				ignoreHTTPSErrors: true
			});
		} else {
			// running on vercel
			return chromium.puppeteer.launch({
				args: chromium.args,
				executablePath,
				headless: chromium.headless,
				ignoreHTTPSErrors: true
			});
		}
	} catch (e) {
		console.log(e);
	}
};

```

Locally everything works fine.

But on vercel i get:
```
{
        "level": "error",
        "message": "Error: Could not find expected browser (chrome) locally. Run `npm install` to download the correct Chromium revision (970485).\n    at ChromeLauncher.launch (/var/task/node_modules/puppeteer/lib/cjs/puppeteer/node/Launcher.js:105:23)\n    at async getGroupPostsByKeywords (/var/task/.next/serverless/pages/api/[endpoint].js:153:15)\n    at async handler (/var/task/.next/serverless/pages/api/[endpoint].js:213:10)\n    at async handleRequest (/var/task/.next/serverless/pages/api/[endpoint].js:236:16)\n    at async api (/var/task/.next/serverless/pages/api/[endpoint].js:263:20)\n    at async apiResolver (/var/task/node_modules/next/dist/next-server/server/api-utils.js:8:1)\n    at async module.exports.Gb4g.__webpack_exports__.default (/var/task/.next/serverless/pages/api/[endpoint].js:617:11)",
        "timestamp": "2022-03-31T11:41:01.329Z"
      }
```

Does anyone know whats going wrong here?
Do i have to set additional env variables on vercel?

**Reply by ndom91** [reply]
Looks like either the `chrome-aws-lambda` install or the `executablePath` thereof isn't resolving corrently or getting picked up correctly. Or maybe `puppeteer-core@13.5.1` is expecting a different version compared to what `chrome-aws-lambda@10.1.10` is delivering?

**Reply by CWarth93** [reply]
after having a look at the supported versions i updated to

```
"dependencies": {
    "chrome-aws-lambda": "10.1.0",
    "dotenv": "^8.2.0",
    "moment": "^2.29.1",
    "next": "10.0.0",
    "prettier": "^2.0.2",
    "puppeteer-core": "10.1.0",
    "react": "^16.13.1",
    "react-youtube": "^7.14.0",
    "riteway": "^6.1.2"
  },
  "devDependencies": {
    "puppeteer": "10.1.0"
  }
```

now i get

`
"Error: Could not find expected browser (chrome) locally. Run `npm install` to download the correct Chromium revision (884014).\n    at ChromeLauncher.launch (/var/task/node_modules/puppeteer/lib/cjs/puppeteer/node/Launcher.js:82:27)\n    at async getGroupPostsByKeywords (/var/task/.next/serverless/pages/api/[endpoint].js:153:15)\n    at async handler (/var/task/.next/serverless/pages/api/[endpoint].js:213:10)\n    at async handleRequest (/var/task/.next/serverless/pages/api/[endpoint].js:236:16)\n    at async api (/var/task/.next/serverless/pages/api/[endpoint].js:263:20)\n    at async apiResolver (/var/task/node_modules/next/dist/next-server/server/api-utils.js:8:1)\n    at async module.exports.ndtY.__webpack_exports__.default (/var/task/.next/serverless/pages/api/[endpoint].js:674:11)"`

884014 is the correct version according to https://www.npmjs.com/package/chrome-aws-lambda

**Comment by vasco3** [comment]
It works if you only use puppeteer + chrome-aws-lambda in that server.
As long as you add more stuff like next.js it won't fit on the instance

**Reply by CWarth93** [reply]
does this mean we cant run it as a serverless function in our api?

i can put it to another build like here https://github.com/whitep4nth3r/puppeteer-demo but i would at least need a cors package to access

**Reply by vasco3** [reply]
seems like it because it is too big so it doesn't fit with any other big dependency like Next.
so I deploy it in a static project with a `api/pdf` route and that works as a server less function

the drawback is that I need to call it using the full url path from my nextJS app `https://blalbla.com/api/pdf` instead of just `/api/pdf`

My guess would be that maybe there is a setting in vercel to increase the memory limit for server less functions but I haven't been successful on that.

so if you just want to get going, I would just put puppeteer on its own instance

**Reply by CWarth93** [reply]
how do you set cors policies in this new static project to access the api? (without installing new packages)

**Reply by vasco3** [reply]
I don't set it because I open it with `window.open`

**Comment by boyphongsakorn** [comment]
`id=24","[pid=24][err] /tmp/chromium: error while loading shared libraries: libnss3.so: cannot open shared object file: No such file or directory"`

um.... not work for me by using puppeteer or Playwright.
how to install libnss or fix this error?

**Reply by ndom91** [reply]
Was getting this error too with `node16`, looks like you have to bump it back down to `node14` in the main project settings.

**Reply by boyphongsakorn** [reply]
> Was getting this error too with `node16`, looks like you have to bump it back down to `node14` in the main project settings.

wait what version of chrome-aws-lambda and puppeteer-core are you using?

i use 2.1.1 but only work in node12

**Comment by vasco3** [comment]
I think that puppeteer doesn't work in node v16

**Comment by ndom91** [comment]
Also oddly enough, I see this output about my `/api/image` function being too big (`50mb+`) in the build logs.

![image](https://user-images.githubusercontent.com/7415984/170332058-439a4940-fba8-4105-896c-8ef15038e2d7.png)


However it still seems to work in that deployment :thinking: 

This is with the latest `puppeteer-core` and `chrome-aws-lambda` btw.

**Reply by jonasgroendahl** [reply]
Got it working using older versions that were smaller in size

 "chrome-aws-lambda": "6.0.0"
 "puppeteer-core": "6.0.0"
 
 Not a great long term solution but works :)

**Reply by ndom91** [reply]
I probably shouldn't have posted this publicly 😅, mine started to fail to build now after a few days. Still ~50.2mb compressed

**Reply by ndom91** [reply]
Little tip: I downgraded `chrome-aws-lambda` from `10.1.0` (Chrome 92) to `9.1.0` (Chrome 91) and that shaved a few mb off the final api route's size. Allowing me to deploy again.

**Reply by vasco3** [reply]
so with a leaner `chrome-aws-lambda` it should work. Or if we could expand the size limit of the lambda

**Reply by ndom91** [reply]
Right. The lambda size is enforced upon Vercel by AWS though unfortunately (See: https://docs.aws.amazon.com/lambda/latest/dg/gettingstarted-limits.html), so theres not much we can do there..

**Comment by egroeg92** [comment]
I'm getting this issue - Vercel is unable to find the path to the browser.  Does anyone know how to fix this??

```
c3487a7b-7bfe-435e-8c78-6723f2da8a39	ERROR	Unhandled Promise Rejection 
{"errorType":"Runtime.UnhandledPromiseRejection","errorMessage":"Error: Cannot find module '/var/task/node_modules/puppeteer-core/lib/cjs/puppeteer/common/Browser'","reason":
{"errorType":"Error","errorMessage":"Cannot find module '/var/task/node_modules/puppeteer-core/lib/cjs/puppeteer/common/Browser'","code":"MODULE_NOT_FOUND","path":"/var/task/node_modules/puppeteer-core/package.json","stack":["Error: Cannot find module '/var/task/node_modules/puppeteer-core/lib/cjs/puppeteer/common/Browser'","    at createEsmNotFoundErr (node:internal/modules/cjs/loader:960:15)","    at finalizeEsmResolution (node:internal/modules/cjs/loader:953:15)","    at resolveExports (node:internal/modules/cjs/loader:482:14)","    at Function.Module._findPath (node:internal/modules/cjs/loader:522:31)","    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:919:27)","    at Function.Module._load (node:internal/modules/cjs/loader:778:27)","    at Module.require (node:internal/modules/cjs/loader:1005:19)","    at require (node:internal/modules/cjs/helpers:102:18)","    at Object.<anonymous> (/var/task/node_modules/chrome-aws-lambda/source/puppeteer/lib/Browser.ts:9:11)","    at Module._compile (node:internal/modules/cjs/loader:1105:14)"]},"promise":{},"stack":["Runtime.UnhandledPromiseRejection: Error: Cannot find module '/var/task/node_modules/puppeteer-core/lib/cjs/puppeteer/common/Browser'","    at process.<anonymous> (file:///var/runtime/index.mjs:775:15)","    at process.emit (node:events:539:35)","    at process.emit (/var/task/___vc/__sourcemap_support.js:587:21)","    at emit (node:internal/process/promises:140:20)","    at processPromiseRejections (node:internal/process/promises:274:27)","    at processTicksAndRejections (node:internal/process/task_queues:97:32)"]}
Unknown application error occurred
```

**Reply by ndom91** [reply]
Are you using `puppeteer-core` and `chrome-aws-lambda` packages?

If so, you should launch puppeteer with these options and it should work (with workaround for local dev):

```js
import chromium from 'chrome-aws-lambda'
import puppeteer from 'puppeteer-core'

const browser = await puppeteer.launch({
  args: chromium.args,
  executablePath:
    process.env.NODE_ENV !== 'development'
      ? await chromium.executablePath
      : '/bin/chromium',
  headless: true,
})
```

**Reply by egroeg92** [reply]
Yes, the following are in my package.json dependencies :
    "chrome-aws-lambda": "^10.1.0",
    "puppeteer-core": "^15.1.0",

I've also moved my api folder into the root folder for vercel (which successfully solved the 50mb problem) .  Puppeteer seems to be working fine locally, but I'm the above error log in the vercel function logs...

this is what my code looks like to get the headless html
```
import chromium from 'chrome-aws-lambda';
import puppeteer from 'puppeteer-core';

export default async function handler(req, res) {
	if (req.method === 'POST') {
		const browser = await puppeteer.launch({
			args: [...chromium.args, '--disable-web-security'],
			defaultViewport: chromium.defaultViewport,
			executablePath: 
            process.env.NODE_ENV !== 'development' ? 
            await chromium.executablePath :
            '/bin/chromium',
			headless: true,
			ignoreHTTPSErrors: true,
		});

		const page = await browser.newPage();
		await page.goto(req.body.pageUrl, { waitUntil: 'load', timeout: 0 });

		const body = await page.evaluate(() => {
			return document.body.parentElement.innerHTML;
		});

		await browser.close();
		res.send({ data: body });
	} else {
		// Handle any other HTTP method
	}
}
```

**Reply by ndom91** [reply]
Hmm strange, looks good to me on the surface. I know I was having problems with Node `16` on Vercel and had to explicitly set it down to Node `14` in the project settings.

Otherwise, I'm just using:

- `puppeteer-core@14.3.0`
- `chrome-aws-lambda@9.1.0`

And everything seemed to work without any other fussing about..

**Reply by egroeg92** [reply]
I tried node 14.x.x and different puppeteer/aws lamdas before (away from my comp right now so not sure the specific version numbers) but I'll try again with these versions and report back. 

You are using node 14.0.0?

**Reply by ndom91** [reply]
Yup, those specific versions of the two packages and node 14 in the Vercel project settings. 

I downgraded a bit to somewhat older versions of `puppeteer-core` and `chrome-aws-lambda` because they were also a good bit smaller and let me slide in under the 50mb limit and still use some other code / libraries as well.

**Reply by egroeg92** [reply]
stilll no luck
![Screen Shot 2022-06-27 at 10 15 57 AM](https://user-images.githubusercontent.com/5962575/175999118-2042204e-da96-4cde-8ead-72e3b0603ccf.png)

**Reply by ndom91** [reply]
> stilll no luck
> ![Screen Shot 2022-06-27 at 10 15 57 AM](https://user-images.githubusercontent.com/5962575/175999118-2042204e-da96-4cde-8ead-72e3b0603ccf.png)
> 
> 

Okay the only other thing I can think of, off the top of my head, is to double and triple check that its really using the `chromium.executablePath`( from `chrome-aws-lambda`) instead of the fallback `/bin/chromium` or whatever you've set it to 🤞

**Reply by egroeg92** [reply]
my executable path is returning as `/tmp/chromium`  (which I guess is the path on vercel)
should it be `build/puppeteer/lib/Browser` ??

**Reply by ndom91** [reply]
Unfortuntaely I'm running into an issue in my project atm where I'm over 50mb on the API route, so I can't deploy and log out my `executablePath` to double check.

**Reply by vasco3** [reply]
I use a vercel project exclusively for puppeteer + `chrome-aws-lamda` and I'm also stuck in Nodejs v14

**Reply by ndom91** [reply]
@egroeg92 so I ended up going the route of just making a separate Next.js app / Vercel project just for the "screenshot API" portion of my application :/

Yes, kind of a pain in the ass, but at least I have that separated off now into its own thing and can add as much as I want to my normal API routes. 

Anyway, on that new project, I initially had the same error as you, couldn't find the chromium binary at `/tmp/chromium`. However, as soon as I updated the project settings down to use `node 14` it worked like a charm :+1: 

Using `chrome-aws-lambda@9.1.0` and `puppeteer-core@15.3.0` this time around.

**Reply by ryoid** [reply]
Managed to get it working by downgrading to node 14 after getting same error as @ndom91 
- `chrome-aws-lambda@10.1.0`
 - `puppeteer-core@15.3.2`

I set the node version in my `package.json` instead of project settings
```json
"engines": {
    "node": "14.x"
 }
```

**Reply by egroeg92** [reply]
I ended up switching to EC2 and the code worked :-( . Didn't try having the headless on a seperate vercel instance though.  Thanks for your help!

**Comment by mityabeldii** [comment]
This is an official solution from Vercel and it works fine for me: https://github.com/vercel/virtual-event-starter-kit/blob/main/lib/screenshot.ts

Only the latest versions of `chrome-aws-lambda` and `puppeteer-core` were installed.
It's also a workaround for the 50mb limitation.

**Reply by ndom91** [reply]
Yeah that might work if yuo don't have any other larger API routes. They all get bundled into one lambda apparently (even though they're supposed to get split after 50mb+, but that still didn't seem to help me). 

So once yuo start adding dependencies to your other API routes, that'll burst over the limit.

**Comment by leerob** [comment]
Hey! If anyone is still trying to use Chromium/Puppeteer for taking screenshots inside Serverless Functions, could you send me an email `lee@vercel.com` with more information about what you're trying to build? Thank you!

**Comment by jgalianoz** [comment]
I am trying to use puppeteer but my deployment fails. I am using remix.

<img width="919" alt="image" src="https://user-images.githubusercontent.com/12010942/180266058-cda9c956-c6e5-4e92-8fff-359262bf0a51.png">

**Comment by Atharv-Chaudhari** [comment]
```
const express = require('express'),
    app = express();
const chromium = require('chrome-aws-lambda');
const playwright = require('playwright-core');

app.get("/", async (request, response) => {
  try {
    const browser = await playwright.chromium.launch({
    args: [...chromium.args, "--hide-scrollbars", "--disable-web-security"],
    defaultViewport: chromium.defaultViewport,
    executablePath: await chromium.executablePath,
    headless: true,
    ignoreHTTPSErrors: true,
  });
```

![image](https://user-images.githubusercontent.com/61587515/184506218-f0adbecb-39d4-4cbf-96a3-9dda4b908b60.png)

Getting this error for playwright on vercel 👀

**Reply by Atharv-Chaudhari** [reply]
I ended up this using following code..

const browser = await chromium.puppeteer.launch({
        args: [...chromium.args, "--hide-scrollbars", "--disable-web-security"],
        defaultViewport: chromium.defaultViewport,
        executablePath: await chromium.executablePath,
        headless: true,
        **ignoreHTTPSErrors: true,**
        ignoreDefaultArgs: ['--disable-extensions'],
    });

**Reply by ndom91** [reply]
Are you using `node16` in that Vercel project? Seems `node14` might be required for puppeteer/playwright

**Reply by Atharv-Chaudhari** [reply]
yes node 14 works

**Comment by ivenuss** [comment]
Any updates or possible solutions?

**Reply by vasco3** [reply]
I think it will only get worse as Chromium keeps getting heavier.
There will have to be a tool that is as lean as possible and that just does PDFs

**Reply by ivenuss** [reply]
I'm looking for a tool that generates an image from HTML and it has to be server sided :/

**Reply by vasco3** [reply]
Vercel is working on a tool for that case @ivenuss 
@leerob is running a beta

**Reply by fahadamjadckj** [reply]
Hi, after spending a day and a half I have finally got a working version with "chrome-aws-lambda": "^8.0.2" and "puppeteer-core": "^8.0.0",
you also need to set engines to be node version to 14 like so in package.json.
{
 ...
  "engines": {
    "node": "^14"
  }
  ...
}

**Reply by ndom91** [reply]
I went the route of making my screenshot api route into its own Next.js / Vercel project. That way I had the full 50mb to play with there, unencumbered by any other potential api route(s) and their dependencies. 

However, that also got too large with the latest playwright-core and chrome-aws-lambda versions. 

So I moved that single api route next.js project to a simple fastify server with 1 endpoint as well. Other than the scaffolding for fastify, I was able to almost 1:1 copy the contents of my playwright next.js api route over. 

This also came with the advantage of better performance as well. Went from generating screenshots of URLs in 7-10s (meaning it would completely fail and time out every once in a while as well) with next.js to 2-5s with fastify

**Reply by fahadamjadckj** [reply]
Where did you host the fastify server? Aws, google cloud?

**Reply by ndom91** [reply]
> Where did you host the fastify server? Aws, google cloud?

In the same Vercel project as previously hosted the next.js app actually haha. The migration was super easy. 

You can check out the fastify project at https://github.com/ndom91/briefkasten-screenshot

**Reply by fahadamjadckj** [reply]
Oh ok ok, I am also facing the same timeout issue becauase data fetching took so long, I will look into it.

**Comment by mehedi-sharif** [comment]
Hello Guys, I am getting an error on vercel:

```
Error: Could not find expected browser (chrome) locally. Run `npm install` to download the correct Chromium revision (901912).
--
09:12:17.130 | at ChromeLauncher.launch (/vercel/path0/node_modules/puppeteer-core/lib/cjs/puppeteer/node/Launcher.js:88:27)
09:12:17.130 | at async screenshot (file:///vercel/path0/src/vercel.js:13:19)
09:12:17.138 | Error: Command "npm run build" exited with 1
```

and this is my code:
```
import chrome from "chrome-aws-lambda";
import puppeteer from "puppeteer-core";

async function screenshot() {
  const options = {
    args: [...chrome.args, "--hide-scrollbars", "--disable-web-security"],
    defaultViewport: chrome.defaultViewport,
    executablePath: await chrome.executablePath,
    headless: true,
    ignoreHTTPSErrors: true,
  };
  console.log("trying to execute");
  const browser = await puppeteer.launch(options);
  console.log("browser launched");
  const page = await browser.newPage();
  await page.setViewport({ width: 2000, height: 1000 });
  await page.goto("https://demo.themefisher.com/bexer/", {
    waitUntil: "networkidle0",
  });
  await page.screenshot({ path: "bexer.png" });

  await browser.close();
}
screenshot();
```

can anyone help with that?

**Reply by fahadamjadckj** [reply]
Use node version 14 and make sure your aws lamda version is compatiable with puppeteer version

**Reply by mehedi-sharif** [reply]
That doesn't solve the issue. thanks

**Comment by kettanaito** [comment]
Hi, folks. If you're still struggling to make this work, I wrote a guide about how I achieved Chromium on Vercel serverless today. 

- [**Chromium on Vercel (serverless)**](https://gist.github.com/kettanaito/56861aff96e6debc575d522dd03e5725) in 2022

It lists the exact steps you need to take, emphasizes the right dependency versions to use (yes, those do matter), and includes solutions to the common problems I (and many of you here) faced while trying to make this work, like `The Serverless Function "XYZ" is XX.YYmb which exceeds the maximum size limit of 50mb` or `Error while loading shared libraries: libnss3.so: cannot open shared object file`.

Build something awesome.
